api/src/main/java/com/microsoft/gctoolkit/jvm/AbstractJavaVirtualMachine.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.Phaser;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * An implementation of JavaVirtualMachine that uses io.vertx verticles to feed
+ * lines to the parser(s) and post events to the aggregators. This implementation
+ * is here in the vertx module so that the api and parser modules can exist without
+ * having to import io.vertx. In the api module, the class GCToolKit uses the classloader
+ * to load the appropriate JavaVirtualMachine.
+ */
+public abstract class AbstractJavaVirtualMachine implements JavaVirtualMachine {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractJavaVirtualMachine.class.getName());
+    private static final double LOG_FRAGMENT_THRESHOLD_SECONDS = 60.0d; //todo: replace magic threshold with a heuristic
+
+    private GCLogFile dataSource;
+    private Diary diary;
+    private DateTimeStamp estimatedStartTime;
+    private DateTimeStamp timeOfLastEvent;
+    private double logDuration = -1.0d;
+    private final Map<Class<? extends Aggregation>, Aggregation> aggregatedData = new ConcurrentHashMap<>();
+
+    public void setDataSource(DataSource logFile) throws IOException {
+        this.dataSource = (GCLogFile) logFile;
+        this.diary = logFile.diary();
+    }
+
+    @Override
+    public boolean isG1GC() {
+        return diary.isG1GC();
+    }
+
+    @Override
+    public boolean isZGC() {
+        return diary.isZGC();
+    }
+
+    @Override
+    public boolean isShenandoah() {
+        return diary.isShenandoah();
+    }
+
+    @Override
+    public boolean isParallel() {
+        return diary.isPSYoung();
+    }
+
+    @Override
+    public boolean isSerial() {
+        return diary.isSerialFull();
+    }
+
+    @Override
+    public boolean isCMS() {
+        return diary.isCMS();
+    }
+
+    @Override
+    public String getCommandLine() {
+        return ""; //todo: extract from diary... jvmConfigurationFromParser.getCommandLine();
+    }
+
+    public DateTimeStamp getTimeOfFirstEvent() {
+        return diary.getTimeOfFirstEvent();
+    }
+
+    /**
+     * todo: fix this to be a globally available value. Also, the JVM start time is not zero if the time
+     * If the first event is significantly distant from zero in relation to the time intervals between the
+     * of the next N events, where N maybe 1, then this is likely a log fragment and not the start of the run.
+     * <p>
+     * Try to estimate the time at which the JVM started. For log fragments, this will be the time
+     * of the first event in the log. Otherwise it will be 0.000 seconds.
+     *
+     * @return DateTimeStamp
+     */
+    @Override
+    public DateTimeStamp getEstimatedJVMStartTime() {
+        DateTimeStamp startTime = diary.getTimeOfFirstEvent();
```

Comment:
Shouldn't this just be `return estimatedStartTime;` ? 

Does this estimated start time and estimated end time match the same from Aggregator/Aggregation? Should be a unit test for this.